src/imitation/algorithms/dagger.py

```diff
@@ -391,30 +450,182 @@ def save_trainer(self) -> Tuple[str, str]:
+                be completed before a dataset aggregation step ends.
+            bc_train_kwargs: Keyword arguments to pass to `BC.train` during the
+                BC update step. If this is None, then `self._default_bc_train_kwargs()`
+                is used.
+        """
+        total_timestep_count = 0
+        round_num = 0
+
+        while total_timestep_count < total_timesteps:
+            collector = self.get_trajectory_collector()
+            round_episode_count = 0
+            round_timestep_count = 0
+
+            obs = collector.reset()
+            ep_rewards = np.zeros([self.venv.num_envs])
+            # TODO(shwang): This while loop end condition causes rollout collection to
+            #   suffer from the same problem that
+            #   `imitation.data.rollout.generate_trajectories` previously had -- a
+            #   bias towards shorter episodes.
+            #   We could probably solve this problem simply by calling
+            #   `generate_trajectories(self.expert_policy, venv=collector)` now that
+            #   we are ported from using `Env` to `VecEnv`?
+            while (
+                round_episode_count < rollout_round_min_episodes
+                and round_timestep_count < rollout_round_min_timesteps
+            ):
+                acts, _ = self.expert_policy.predict(obs, deterministic=True)
+                obs, rews, dones, _ = collector.step(acts)
+                total_timestep_count += self.venv.num_envs
+                round_timestep_count += self.venv.num_envs
+                ep_rewards += rews
+
+                for i, done in enumerate(dones):
+                    if done:
+                        logger.record_mean("dagger/mean_episode_reward", ep_rewards[i])
+                        ep_rewards[i] = 0
+                round_episode_count += 1
+
+            # Flush partial trajectories so that any timesteps before episode
+            # completion are also available in the dataset.
+            # Without this call, self.extend_and_update() currently
+            # fails if the "round" also didn't finish any episodes (not enough
+            # timesteps).
+            collector.flush_trajectories()
+
+            logger.record("dagger/total_timesteps", total_timestep_count)
+            logger.record("dagger/round_num", round_num)
+            logger.record("dagger/round_episode_count", round_episode_count)
+            logger.record("dagger/round_timestep_count", round_timestep_count)
+
+            # TODO(shwang): It looks like BC might start looping Tensorboard
+            #   back to x=0 with each new call to BC.train(). Consider adding a
+            #   `reset_tensorboard: bool = False` argument to BC.train() if this turns
+            #   out to be the case?
+
+            # `logger.dump` is called inside BC.train within the following fn call:
+            self.extend_and_update(bc_train_kwargs)
```

Comment:
I was initially confused this wasn't passing the data to `extend_and_update`. I now see we save the data to disk, then load it again. What's the point of this indirection? Checkpointing? Demonstrations bigger than RAM?

Reply:
Question for @qxcv

Reply:
When you don't want to keep the process running while you wait for user input (e.g. because it could take several hours for the user to come back, and the run might last for a few days). So I guess checkpointing.

Reply:
Context: I think my motivating application for this code was interactive IL with MTurk demonstrators. For that purpose it would be weird to need a background process running & communicate with IPC—instead you could just save some stuff to disk (or a DB or whatever) as requests come in, then periodically run a cron job to do incremental training.